gtk/src/file_filter_info.rs

```diff
@@ -0,0 +1,71 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use glib::translate::*;
+use libc::c_char;
+
+use crate::FileFilterFlags;
+
+#[derive(Clone)]
+pub struct FileFilterInfo {
```

Comment:
Another option would be to make this a `#[repr(transparent)]` struct around the FFI type and add a) a constructor, and b) getter functions that simply borrow the strings as `&str`.

That way you would reduce the number of copies here.

Reply:
Yeah, that's a good point, would probably be much simpler too.

Reply:
Having a little trouble here.  For the constructor:
```rust
    pub fn new(
        filename: Option<&str>,
        uri: Option<&str>,
        display_name: Option<&str>,
        mime_type: Option<&str>,
    ) -> Self {
        let mut contains = FileFilterFlags::empty();
        for (value, bit) in [
            (filename, FileFilterFlags::FILENAME),
            (uri, FileFilterFlags::URI),
            (display_name, FileFilterFlags::DISPLAY_NAME),
            (mime_type, FileFilterFlags::MIME_TYPE),
        ] {
            if value.is_some() {
                contains |= bit;
            }
        }

        Self(ffi::GtkFileFilterInfo {
            contains: contains.bits(),
            filename: filename.to_glib_full(),
            uri: uri.to_glib_full(),
            display_name: display_name.to_glib_full(),
            mime_type: mime_type.to_glib_full(),
        })
    }
```
For the struct members, if I use `filename.to_glib_none().0`, I end up with gibberish in them later (which makes sense).  But I'm pretty sure `.to_glib_full()` will cause a memory leak (valgrind seems to agree with me).  My first thought was to implement `Drop` and free them there, but I feel like that would cause problems for any `ffi::GtkFileFilterInfo` structs passed to us from C, which would expect to own those `gchar *` pointers, right?  Or do we end up owning anything that comes through that way?

Sorry, I think I just haven't quite wrapped my head around how ownership is supposed to work.

Reply:
Are all those nullable? feels weird

Reply:
Also, for the getters, there seems to be a `from_glib_none()` impl for `Option<String>`, but not for `Option<&str>`.  Am I missing something?  Would be annoying to have to build a `String` for those getters.

Reply:
> Are all those nullable? feels weird

Yes.  The `contains` flags field indicates which are non-null.  I think the idea is that custom filters might not need all fields, so it's more efficient for filters to say what they need, and then for implementations to only provide what's needed.

Reply:
You would use `to_glib_full()` for the strings in the constructor and implement `Drop`. When receiving such a struct from C in case of `FromGlibPtrNone` you would have to `strdup()` all the strings so that `Drop` can free them again.

However as you have a `#[repr(transparent)]` struct you can also borrow the values from C, e.g. in that callback. You can cast a `*const ffi::GtkFileFilterInfo` to `&FileFilterInfo` then via `&*(ptr as *const FileFilterInfo)`.

Reply:
Ah, that makes sense.  I was also looking at how `glib::GString` is implemented, and it looks like it handles the case where it was originally a C object vs. originally a Rust object, and can then presumably be a bit more efficient.  Do you think that's worth it for this?  `GtkFileFilterInfo` structs do seem to get created and dropped in large amounts since often one will get created for each row in a tree model.

(I've realized that `GtkFileFilterInfo` is anunusual case... gtk/glib has very few structs that aren't also GObjects, or aren't at _least_ registered as glib boxed types.  And it appears that none of the other few examples of this have been wrapped yet.  So I feel a little less bad about being fairly clueless here...)

Reply:
Ok, I think I have something that works (pushed), though I don't know if I can reduce string copying further.

One thing I'm having trouble with is that it'd probably be nicer to return borrowed `&str`s from the getters, rather than creating `String`s each time they're called, but I can't seem to figure out how to do that.  For example, I have:

```rust
    pub fn filename(&self) -> Option<String> {
        unsafe { from_glib_none(self.0.filename) }
    }
```

But it doesn't seem like there's an impl of `from_glib_none()` for `&str`, only `String`.  Maybe this isn't really feasible to do, though, since `&str` is fundamentally a borrowed type, and it's not particularly meaningful to create and return those on the fly.  I guess we could return them as `Option<CStr>` without needing any allocation, but maybe that's not worth it, as `CStr` is a bit more of a pain to work with.

Reply:
Argh, also I'm now confused about the `FromGlibPtrBorrow` impl (the generated code appears to need it).  I'm just doing simply `Borrowed::new(FileFilterInfo(*ptr))`, but now won't that cause the `Drop` impl to free someone else's strings when the `Borrowed` (and thus the `FilteFilterInfo`) wrappers get dropped?

Reply:
> [...] and it looks like it handles the case where it was originally a C object vs. originally a Rust object, and can then presumably be a bit more efficient.

I don't think that makes much difference here. When creating one from the Rust side, you will always pass it to C (requires copy because of the strings), whenever you get it from C you usually just borrow it and read the fields but that's it.

So having the constructor create a valid C object (`ffi::GtkFileFilterInfo` with owned C strings that are freed on `Drop`) plus passing a `&FileFilterInfo` into the callback seems the most efficient way here unless I'm missing something.

If you make the `FileFilterInfo` not just a `#[repr(transparent)]` around the FFI struct then you also can't do that cast-and-borrow trick for the callback FWIW.

> But it doesn't seem like there's an impl of from_glib_none() for &str, only String.

You would implement the string getters around `std::ffi::CStr` as otherwise they would still involve a copy of the string. You can find some example of that in the repository :)

Reply:
> You would implement the string getters around std::ffi::CStr as otherwise they would still involve a copy of the string.

Ok, done!  One thing I wasn't sure of was whether to `.expect()` on the result, so that if the string wasn't valid utf-8, it would panic.  I ended up settling on `.ok()` instead, as the fields are nullable and the getters have to return `Option` anyway.  I think I would personally prefer my program not crash due to some invalid data, but, on the other hand, this also silently ignores invalid data, which isn't super great either.

Reply:
I would `expect()` here for invalid UTF-8 but of course `NULL` pointers are allowed and should be mapped to `None`. If we somehow get invalid input then that's a big problem and needs to be solved instead of being hidden, just like we do elsewhere for invalid UTF-8.